File: haskell-server/src/BoardGame/Common/Domain/PieceProviderType.hs
--
-- Copyright 2017-2018 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}

module BoardGame.Common.Domain.PieceProviderType(
  PieceProviderType(..)
  )
  where

import GHC.Generics (Generic)
import Data.Aeson (FromJSON, ToJSON)
import Control.DeepSeq (NFData)

-- | Different types of piece generators.
data PieceProviderType = Random | Cyclic
  deriving (Eq, Show, Generic, NFData)

instance FromJSON PieceProviderType
instance ToJSON PieceProviderType

File: scala-server/app/com/bolour/boardgame/scala/common/domain/PieceProviderType.scala
/*
 * Copyright 2017-2018 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
 */
package com.bolour.boardgame.scala.common.domain

object PieceProviderType {

  val RandomAsString = "Random"
  val CyclicAsString = "Cyclic"

  sealed abstract class PieceProviderType
  object Random extends PieceProviderType {
    override def toString = RandomAsString
  }
  object Cyclic extends PieceProviderType {
    override def toString = CyclicAsString
  }

  def fromString(asString: String): PieceProviderType = {
    asString match {
      case RandomAsString => PieceProviderType.Random
      case CyclicAsString => PieceProviderType.Cyclic
      case _ => throw new RuntimeException(s"unrecognized piece provider type: ${asString}")
    }
  }
}

//object PieceProviderType extends Enumeration {
//  type PieceProviderType = Value
//  val Random, Cyclic = Value
//}
